crates/bevy_ecs/src/system/system_registry.rs

```diff
@@ -0,0 +1,645 @@
+/// struct Counter(u8);
+///
+/// fn count_up(mut counter: ResMut<Counter>){
+///     counter.0 += 1;
+/// }
+///
+/// world.init_resource::<Counter>();
+/// world.run_system(count_up);
+///
+/// assert_eq!(Counter(1), *world.resource());
+/// ```
+///
+/// These systems immediately apply commands and cache state,
+/// ensuring that change detection and [`Local`](crate::system::Local) variables work correctly.
+///
+/// ```rust
+/// use bevy_ecs::prelude::*;
+///
+/// let mut world = World::new();
+///
+/// #[derive(Component)]
+/// struct Marker;
+///
+/// fn spawn_7_entities(mut commands: Commands) {
+///     for _ in 0..7 {
+///         commands.spawn().insert(Marker);
+///     }
+/// }
+///
+/// fn assert_7_spawned(query: Query<(), Added<Marker>>){
+///     let n_spawned = query.iter().count();
+///     assert_eq!(n_spawned, 7);
+/// }
+///
+/// world.run_system(spawn_7_entities);
+/// world.run_system(assert_7_spawned);
+/// ```
+#[derive(Default)]
+pub struct SystemRegistry {
+    systems: Vec<StoredSystem>,
+    // Stores the index of all systems that match the key's label
+    labels: HashMap<Box<dyn SystemLabel>, Vec<usize>>,
+}
+
+struct StoredSystem {
+    system: Box<dyn System<In = (), Out = ()>>,
+}
+
+impl SystemRegistry {
+    /// Registers a system in the [`SystemRegistry`], so then it can be later run.
+    ///
+    /// This allows the system to be run by their [`SystemTypeIdLabel`] using the `run_systems_by_label` method.
+    /// Repeatedly registering a system will have no effect.
+    ///
+    /// Ordinarily, systems are automatically registered when [`run_system`](SystemRegistry::run_system) is called.
+    /// When this occurs, they will be registered using their [`SystemTypeIdLabel`].
+    /// However, manual registration allows you to provide one or more labels for the system.
+    /// This also allows you to register multiple distinct copies of the system under distinct labels.
+    ///
+    /// When [`run_systems_by_label`](SystemRegistry::run_systems_by_label) is called,
+    /// all registered systems that match that label will be evaluated (in insertion order).
+    ///
+    /// To provide explicit label(s), use [`register_system_with_labels`](SystemRegistry::register_system_with_labels).
+    #[inline]
+    pub fn register_system<Params, S: IntoSystem<(), (), Params> + 'static>(
+        &mut self,
+        world: &mut World,
+        system: S,
+    ) {
+        let boxed_system: Box<dyn System<In = (), Out = ()>> =
+            Box::new(IntoSystem::into_system(system));
+        let automatic_system_label: SystemTypeIdLabel<S> = SystemTypeIdLabel::new();
+
+        // This avoids nasty surprising behavior in case systems are registered twice
+        if !self.is_label_registered(automatic_system_label) {
+            self.register_boxed_system_with_labels(
+                world,
+                boxed_system,
+                vec![Box::new(automatic_system_label)],
+            );
```

Comment:
```suggestion
        let automatic_system_label: SystemTypeIdLabel<S> = SystemTypeIdLabel::new();

        // This avoids nasty surprising behavior in case systems are registered twice
        if !self.is_label_registered(automatic_system_label) {
        	let boxed_system: Box<dyn System<In = (), Out = ()>> =
            	Box::new(IntoSystem::into_system(system));
            self.register_boxed_system_with_labels(
                world,
                boxed_system,
                vec![Box::new(automatic_system_label)],
            );
```
This avoids an Box allocation, if system was already registered.